File: coding/src/com/algos/topics/li04_twopointers/revised/FindtheIntersectionOfTwoSortedArrays.java
package com.algos.topics.li04_twopointers.revised;

import java.util.ArrayList;
import java.util.List;

public class FindtheIntersectionOfTwoSortedArrays {

        public static void main(String[] args) {
            FindtheIntersectionOfTwoSortedArrays obj = new FindtheIntersectionOfTwoSortedArrays();
            int[] nums1 = { 1, 2, 3, 4, 5 };
            int[] nums2 = { 1, 2, 3 };
            System.out.println(obj.findIntersection(nums1, nums2));
        }

        // Problem: Find the intersection of two sorted arrays
        public List<Integer> findIntersection(int[] nums1, int[] nums2){
            int n1 = nums1.length, n2 = nums2.length;
            int i=0, j=0;
            List<Integer> intersection = new ArrayList<>();
            while(i<n1 && j<n2){
                if(nums1[i] == nums2[j]){
                    intersection.add(nums1[i]);
                    i++;
                    j++;
                }else if(nums1[i] < nums2[j]){
                    i++;
                }else{
                    j++;
                }
            }

            return intersection;
        }
}

File: dsa/topics/03_twopointers/python/find_the_intersection_of_two_sorted_arrays.py
# Find the Intersection of Two Sorted Arrays
#
# Problem: Find the intersection of two sorted arrays
#
# Idea:
# - Use two pointers to traverse both arrays
# - Move the pointer of the smaller element
#
# Time: O(n + m)
# Space: O(1) excluding the result array


def find_intersection(nums1, nums2):
    i = j = 0
    intersection = []

    while i < len(nums1) and j < len(nums2):
        if nums1[i] == nums2[j]:
            intersection.append(nums1[i])
            i += 1
            j += 1
        elif nums1[i] < nums2[j]:
            i += 1
        else:
            j += 1

    return intersection


# -------------------- TEST --------------------
if __name__ == "__main__":
    nums1 = [1, 2, 3, 4, 5]
    nums2 = [1, 2, 3]
    print(find_intersection(nums1, nums2))  # Output: [1, 2, 3]
